packages/cid-verifier/src/verification.js

```diff
@@ -0,0 +1,130 @@
+ * @param {(params: Array<string>, request: Request, env: import('./env').Env) => Promise<Response>} fn
+ * @returns {import('itty-router').RouteHandler<Request>}
+ */
+function withRequiredQueryParams (params, fn) {
+  /**
+   * @param {Request} request
+   * @param {import('./env').Env} env
+   */
+  return async function (request, env) {
+    const searchParams = (new URL(request.url)).searchParams
+
+    for (const param of params) {
+      if (!searchParams.get(param)) {
+        return new Response(`${param} is a required query param`, { status: 400 })
+      }
+    }
+
+    return await fn(params.map(param => String(searchParams.get(param))), request, env)
+  }
+}
+
+export const verificationGet = withRequiredQueryParams(['cid'],
+  /**
+   * Returns google malware result.
+   */
+  async function (params, request, env) {
+    const [cid] = params
+
+    const denyListResource = await getFromDenyList(cid, env)
+    if (denyListResource) {
+      const { status } = JSON.parse(denyListResource)
+      if (status === 451) {
+        return new Response('BLOCKED FOR LEGAL REASONS', { status: 451 })
+      } else {
+        return new Response('MALWARE DETECTED', { status: 403 })
+      }
+    }
+
+    const googleMalwareResults = await getGoogleMalwareResult(cid, env)
+    if (!googleMalwareResults) {
+      return new Response('Verification Not Known', { status: 404 })
+    }
+
+    if (googleMalwareResults === '"PENDING"') {
+      return new Response('Verification Pending', { status: 202 })
+    }
+
+    const parsedGoogleMalwareResults = JSON.parse(googleMalwareResults)
+    if (!Object.keys(parsedGoogleMalwareResults).length) {
+      return new Response('Malware Not Detected', { status: 200 })
+    } else {
+      return new Response('MALWARE DETECTED', { status: 403 })
+    }
+  }
+)
+
+/**
+ * Process CID with malware verification parties.
+ */
+export const verificationPost = withRequiredQueryParams(['cid', 'url'],
+  async function verificationPost (params, request, env) {
+    const [cid, url] = params
+    const googleMalwareResults = await getGoogleMalwareResult(cid, env)
+
+    if (googleMalwareResults === '"PENDING"') {
+      return new Response('Verification Pending', { status: 202 })
+    }
+
+    if (googleMalwareResults) {
+      return new Response('cid already processed', { status: 409 })
+    }
+
+    // cid not processed so we process it
+
+    // we set the status to pending to limit the number of times we call google
+    env.GOOGLE_MALWARE_RESULTS.put(cid, '"PENDING"')
+
+    const googleCloudResponse = await fetch(
+      `${
+        env.GOOGLE_CLOUD_API_URL
+      }/v1/uris:search?threatTypes=MALWARE&threatTypes=SOCIAL_ENGINEERING&threatTypes=UNWANTED_SOFTWARE&uri=${encodeURIComponent(
+        url
+      )}&key=${env.GOOGLE_CLOUD_API_KEY}`
+    )
+
+    if (googleCloudResponse.status === 200) {
+      try {
+        env.GOOGLE_MALWARE_RESULTS.put(cid, await googleCloudResponse.text())
```

Comment:
Consider adding third parameter with `{metadata: ...}` as previously mentioned.

so here in POST, if this is considered malware I think it is easier to just do a new PUT to env.DENYLIST.put

This way, GET only needs to check the denylist because it only cares about if it is bad or not. We internally want to have more information we can share, and that's why it is also useful to have the extra KV.

So, in Denylist we can simply put value { status, reason } like we have in scripts/denylist.js

What do you think?

Reply:
i kinda think it's safer to seperate google and badbits into seperate KV stores.  for any given cid we know exactly what is stored in the respective KVs.

Reply:
Do you mean that you agree with the suggestion, or do you have something else in mind?